### ISSUE TITLE
Automatically clean up OneDrive file versions

### ISSUE THREAD

**Issue by Zvezdin** [rclone/rclone#4106]
#### What is your current rclone version (output from `rclone version`)?
1.51.0

#### What problem are you are trying to solve?
[It is documented](https://rclone.org/onedrive/) that onedrive creates multiple versions of the files uploaded through rclone, using double or even more of the storage quota. The rclone docs mention that this is only a problem with `rclone copy`, however I just did an `rclone sync` to upload ~30k files to onedrive and every document has at least two versions, separately taking storage quota. 

It is not possible to disable file versioning unless you are a SharePoint admin, and Microsoft only provides a way to remove the versions file-by-file.

#### How do you think rclone should be changed to solve that?

The SharePoint .NET API allows the removal of older versions, as [in this project](https://github.com/balassy/OneDriveVersionCleaner). However I could not get this to compile under dotnet core in Linux, possibly because [the specific SharePoint library](https://docs.microsoft.com/en-us/sharepoint/dev/sp-add-ins/complete-basic-operations-using-sharepoint-client-library-code) requires additional Windows dll-s. Nevertheless I think that there might be a way to interface with this API through rclone.

**Comment by Dibbyo456** [comment]
> It is not possible to disable file versioning unless you are a SharePoint admin

Actually it's not possible to disable versioning even if you're sharepoint/office 365 admin. The `EnableMinimumVersionRequirement` parameter has no affect after 2019.

**Comment by ncw** [comment]
You can list the versions of a file here

https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_list_versions?view=odsp-graph-online

I couldn't find a way of deleting previous versions though - can anyone see one?

I suspect to delete versions then you'd have to do a rename or a copy (something which gets rid of the versions) then rename it back. How does that .NET code do it? Can you see?

**Comment by Zvezdin** [comment]
Take a look here: https://github.com/balassy/OneDriveVersionCleaner/blob/master/OneDriveVersionCleaner/Program.cs#L124

I am not sure if this is the documentation for the same API, but here is more clarificaion on the method in question: 
https://docs.microsoft.com/en-us/previous-versions/office/sharepoint-server/ee538326(v=office.15)

**Comment by x0b** [comment]
> I am not sure if this is the documentation for the same API, but here is more clarificaion on the method in question:
https://docs.microsoft.com/en-us/previous-versions/office/sharepoint-server/ee538326(v=office.15)

The C# API cannot be used in rclone.

I performed a few tests and discovered that it is possible to delete versions from Onedrive Business. After obtaining a [ListItemVersion](https://docs.microsoft.com/en-us/graph/api/resources/listitemversion?view=graph-rest-1.0), you can send a ```DELETE /v1.0/sites/{site-id}/items/{item-id}/versions/{version-id}```, which will complete with a 200 OK. The old version is then also immediately gone when verifying using the web UI.

Trying the same on a [DriveItemVersion](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/resources/driveitemversion?view=odsp-graph-online) does lead to a 400 Bad Request though.

**Comment by Saoneth** [comment]
@x0b DELETE for [DriveItemVersion](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/resources/driveitemversion?view=odsp-graph-online) probably didn't work for you, because documentation provides wrong path. Correct one is: `/drives/{drive-id}/items/{item-id}/versions/{version-id}`

I wrote simple script which uses rclone, curl and jq to clean up all old versions:
[onedrive-cleanup.sh](https://github.com/rclone/rclone/files/4838990/onedrive-cleanup.txt)

**Comment by Dibbyo456** [comment]
> I wrote simple script which uses rclone, curl and jq to clean up all old versions: [onedrive-cleanup.sh](https://github.com/rclone/rclone/files/4838990/onedrive-cleanup.txt)

You are a saint my dear. The script works like magic and the best part is OneDrive deletes it permanently (skipping trash).
There was a typo on line 6 though. I hope someone would be kind enough to implement this on Rclone.

**Comment by Dibbyo456** [comment]
@Cnly What do you think about this?

**Comment by Cnly** [comment]
@Dibbyo456 Thanks for pinging. I think this is great since we finally can control the versioning. Would you like to take on this and open a PR?

But speaking of multiple versions created by rclone, IIRC rclone doesn't set the modtime afterwards now as @ncw described (except for [one trivial point](https://github.com/rclone/rclone/issues/2520#issuecomment-418427375)), so the extra versions _should_ only come from overwriting existing files... But things might have changed over time, so some more checking would be good. @Zvezdin Are you seeing multiple versions even for new files, or just for overwritten files?

Apart from that, a `cleanup` command would be good to have anyway.

**Comment by ncw** [comment]
@Cnly I don't know if you've see https://github.com/rclone/rclone/issues/4373 yet but using onedrive with chunker seems to be creating versions

**Comment by Cnly** [comment]
@ncw But if multiple versions are created by just uploading files (no existing ones overwritten) to a non-nested onedrive backend through `rclone copy`, it doesn't seem to be the same case as #4373...

Anyway, perhaps we should follow all metadata-changing operations with a version cleanup.

**Comment by ncw** [comment]
> Anyway, perhaps we should follow all metadata-changing operations with a version cleanup.

Implementing `rclone cleanup` will go a long way towards making happier users! Cleaning up versions for setmodtime would great too.

**Comment by Zvezdin** [comment]
Slightly altered version of @Saoneth 's script to refresh the access token on every folder if you're cleaning up a really large drive + more verbosity: https://pastebin.com/Vh0qnzFk

**Comment by ncw** [comment]
I had a go at the first part of this `rclone cleanup` - testing appreciated!  Run `rclone -i cleanup onedrive:path` and it will delete all old versions in "path". Use the `-i` flag to see what it would do interactively.

This is pretty much a straight port of the above script, except rclone will do the version checking `--checkers` at a time in parallel. The token refresh should be automatic also.

This has exposed a primitive which can remove the versions from a file which we can stick into places if we want, like set mod time but I haven't done that yet.

[v1.52.2-282-g65479e6c-fix-4106-onedrive-cleanup-beta](https://beta.rclone.org/branch/v1.52.2-282-g65479e6c-fix-4106-onedrive-cleanup-beta/) on branch [fix-4106-onedrive-cleanup](https://github.com/rclone/rclone/tree/fix-4106-onedrive-cleanup) (uploaded in 15-30 mins)

**Comment by Dibbyo456** [comment]
Hey nick, thank you so much taking the time to implement this into Rclone. I've tested and works as expected. 👍

```
2020/08/04 08:53:51 DEBUG : Using config file from "C:\\Users\\HARRY\\mir1.conf"
2020/08/04 08:53:54 INFO  : testfile.zip: removing version "2.0"
2020/08/04 08:53:55 INFO  : testfile.zip: removing version "1.0"
2020/08/04 08:53:55 DEBUG : 5 go routines active
```

### The issue:
For me it still does not quite solve the problem yet. I have multiple OneDrive with 25TB storage and most of them are 99% full (10-12 GB free most of them). When doing "sync/copy" to OneDrive and source have multiple files changes the storage run out of space causing the "sync" to fail. I need to run "cleanup" to free up storage then re-run the "sync" again and again. You get the picture.

The solution in my mind is that OneDrive needs a special flag which will delete old versions automatically when a file is updated during sync/copy. Something like `--onedrive-keep-versions` (true|false). This I won't have to run "cleanup" and less chance of running out of space.

**Comment by ncw** [comment]
> Hey nick, thank you so much taking the time to implement this into Rclone. I've tested and works as expected. 

Great!

> For me it still does not quite solve the problem yet. I have multiple OneDrive with 25TB storage and most of them are 99% full (10-12 GB free most of them). When doing "sync/copy" to OneDrive and source have multiple files changes, the storage run out of space (because of versions) causing the "sync" to fail. I need to run "cleanup" to free up storage then re-run the "sync" again and again. You get the picture.

Sure, I knew the above would only be the first step.

> The solution in my mind is that OneDrive needs a special flag which will delete old versions automatically when a file is updated during sync/copy. Something like `--onedrive-keep-versions` (true|false). This I won't have to run "cleanup" and less chance of running out of space. Not just for "set-mod-time" but every updates.

The next part needs a bit of discussion...

Here are the options

1. Make an option `no_versions` which when set will cause to delete all but the latest version. It would do this when uploading files or when doing set mod time.
2. When running set mod time check to see if we made a new version and if so delete the previous one, making set mod time not produce an extra version

I think we could do 1 or 2 or both. 1 is really easy now. 2 is a little bit more complicated and is working around a bug in onedrive which may not always be present so will need to detect that.

I don't think `no_versions` should be the default - I don't want to delete versions when people want to keep them.

Thoughts?

**Comment by Dibbyo456** [comment]
> I think we could do 1 or 2 or both. 1 is really easy now. 2 is a little bit more complicated and is working around a bug in onedrive which may not always be present so will need to detect that.

size diff and mod time diff both causes new versions, so implementing on both would be best IMO.

> I don't think no_versions should be the default - I don't want to delete versions when people want to keep them.

I agree. False should be default.

**Comment by Cnly** [comment]
> I don't think no_versions should be the default - I don't want to delete versions when people want to keep them.

From a network traffic perspective, immediately cleaning up versions after any changes we make will come at a cost of 2 additional requests/file, which slows things down and may also trigger API throttling faster. So I'd make that opt-in.

**Comment by ncw** [comment]
Of those two options no_versions is the simplest. Shall we go with that first?

**Comment by Dibbyo456** [comment]
Yeah. I guess so.

**Comment by ncw** [comment]
OK have a go with this - it makes a new config parameter `no_versions` or `--onedrive-no-versions` which removes versions on update of files or setting the modtime.

[v1.52.2-287-g5ff44c91-fix-4106-onedrive-cleanup-beta](https://beta.rclone.org/branch/v1.52.2-287-g5ff44c91-fix-4106-onedrive-cleanup-beta/) on branch [fix-4106-onedrive-cleanup](https://github.com/rclone/rclone/tree/fix-4106-onedrive-cleanup) (uploaded in 15-30 mins)

onedrive personal creates versions but the API to delete them doesn't work. In fact this API isn't mentioned in [the docs](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_list_versions?view=odsp-graph-online) at all as far as I can see! The API for listing versions works on Personal and Business but deleting them only works on Business. Any ideas @Cnly @x0b ?

**Comment by Dibbyo456** [comment]
Outstanding update. I've tested with "size diff" and "mod-time diff" and it worked perfectly for both cases.

**Size diff**:
```
2020/08/07 06:41:56 DEBUG : testfile.rar: Sizes differ (src 11503183 vs dst 11503464)
2020/08/07 06:41:56 DEBUG : testfile.rar: Starting multipart upload
2020/08/07 06:41:56 DEBUG : testfile.rar: Uploading segment 0/11503183 size 10485760
2020/08/07 06:41:59 DEBUG : testfile.rar: Uploading segment 10485760/11503183 size 1017423
2020/08/07 06:42:03 INFO  : testfile.rar: removing version "2.0"
2020/08/07 06:42:03 INFO  : testfile.rar: removing version "1.0"
2020/08/07 06:42:04 DEBUG : testfile.rar: QuickXorHash = 835124a94754baa3c1304d78eab76a0b6c2d0885 OK
2020/08/07 06:42:04 INFO  : testfile.rar: Copied (replaced existing)
```

**Mod-time diff**:
```
2020/08/07 06:42:27 DEBUG : testfile.rar: Modification times differ by -34.8047185s: 2020-08-07 06:42:19.8047185 +0530 IST, 2020-08-07 01:11:45 +0000 UTC
2020/08/07 06:42:27 DEBUG : testfile.rar: QuickXorHash = 835124a94754baa3c1304d78eab76a0b6c2d0885 OK
2020/08/07 06:42:28 INFO  : testfile.rar: removing version "3.0"
2020/08/07 06:42:28 INFO  : testfile.rar: Updated modification time in destination
2020/08/07 06:42:28 DEBUG : testfile.rar: Unchanged skipping
```

But one question though, shouldn't Rclone print `removing version` line after the `Copied (replaced existing)` or `Updated modification time in destination` line? It's nothing just wanted to ask.

**Comment by Cnly** [comment]
@ncw I did some testing and it seems versions don't use extra space on OneDrive Personal. Their web interface also doesn't have a delete button for versions. So I guess we're good to ignore them?

**Comment by ncw** [comment]
@Dibbyo456 wrote:
> Outstanding update. I've tested with "size diff" and "mod-time diff" and it worked perfectly for both cases.

Great!

> But one question though, shouldn't Rclone print `removing version` line after the `Copied (replaced existing)` or `Updated modification time in destination` line? It's nothing just wanted to ask.

The onedrive backend doesn't remove the versions until after the copy is successful - it does this just before returning to the higer level code doing the copy which is the bit that prints `Copied (replaced existing)`. So I don't think I can easily make the logs the other way round, but I don't think it is anything to worry about as we know the version is safely uploaded before deleting the old ones.

I guess the post copy checks could fail (eg checksum) in which case we've deleted the old version before the new one is good, however this is quite unlikely and rclone will immediately retry.

@Cnly wrote
> I did some testing and it seems versions don't use extra space on OneDrive Personal.

That is good

> Their web interface also doesn't have a delete button for versions. So I guess we're good to ignore them?

I've suggested in the docs for `rclone cleanup` and `no_versions` that they don't work on Onedrive Personal. Users can turn them on if it is ever needed. I don't think it is necessary to disable these features from working on Onedrive Personal do you @Cnly?

I've merged where we've got to now to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.53

**Comment by Cnly** [comment]
> I've suggested in the docs for rclone cleanup and no_versions that they don't work on Onedrive Personal. Users can turn them on if it is ever needed. I don't think it is necessary to disable these features from working on Onedrive Personal do you @Cnly?

Actually I do think we should skip that for OneDrive Personal now or at some point, given the performance penalty and probably some misconception people will have like "I should always clean the versions on my OneDrive Personal." It's not a huge problem though - maybe many using Personal just won't notice the versioning at all. So I'm okay with what we have now, and I can follow this up later if needed.
